components/TextFields/src/MDCTextInputBorderView.m

```diff
@@ -0,0 +1,171 @@
+    case kCGLineJoinMiter:
+      lineJoinString = kCALineJoinMiter;
+      break;
+    case kCGLineJoinRound:
+      lineJoinString = kCALineJoinRound;
+      break;
+    default:
+      break;
+  }
+  return lineJoinString;
+}
+
+@interface MDCTextInputBorderView()
+
+@property(nonatomic, strong, readonly) CAShapeLayer *borderLayer;
+
+@end
+
+@implementation MDCTextInputBorderView
+
+- (instancetype)initWithFrame:(CGRect)frame
+{
+  self = [super initWithFrame:frame];
+  if (self) {
+    [self commonMDCTextInputBorderViewInit];
+  }
+  return self;
+}
+
+- (instancetype)initWithCoder:(NSCoder *)coder
+{
+  self = [super initWithCoder:coder];
+  if (self) {
+    _borderFillColor = [coder decodeObjectForKey:MDCTextInputBorderViewBorderFillColorKey];
+    _borderPath = [coder decodeObjectForKey:MDCTextInputBorderViewBorderPathKey];
+    _borderStrokeColor = [coder decodeObjectForKey:MDCTextInputBorderViewBorderStrokeColorKey];
+    [self commonMDCTextInputBorderViewInit];
+  }
+  return self;
+}
+
+- (void)encodeWithCoder:(NSCoder *)aCoder {
+  [super encodeWithCoder:aCoder];
+  [aCoder encodeObject:self.borderFillColor forKey:MDCTextInputBorderViewBorderFillColorKey];
+  [aCoder encodeObject:self.borderPath forKey:MDCTextInputBorderViewBorderPathKey];
+  [aCoder encodeObject:self.borderStrokeColor forKey:MDCTextInputBorderViewBorderStrokeColorKey];
+}
+
+- (nonnull id)copyWithZone:(nullable NSZone *)zone {
+  MDCTextInputBorderView *copy = [[[self class] alloc] initWithFrame:self.frame];
+  copy.borderFillColor = self.borderFillColor;
+  copy.borderPath = self.borderPath.copy;
+  copy.borderStrokeColor = self.borderStrokeColor;
+
+  return copy;
+}
+
+- (void)commonMDCTextInputBorderViewInit {
+  _borderFillColor = _borderFillColor ? _borderFillColor : [UIColor clearColor];
+  _borderStrokeColor = _borderStrokeColor ? _borderStrokeColor : [UIColor clearColor];
+
+  self.userInteractionEnabled = NO;
+  self.opaque = NO;
+
+  self.borderLayer.backgroundColor = [UIColor clearColor].CGColor;
+  self.borderLayer.contentsScale = UIScreen.mainScreen.scale;
+  self.borderLayer.opaque = NO;
+  self.borderLayer.rasterizationScale = self.borderLayer.contentsScale;
+  self.borderLayer.shouldRasterize = YES;
+  self.borderLayer.zPosition = -1.f;
+}
+
+- (void)updateBorder {
+  self.borderLayer.fillColor = self.borderFillColor.CGColor;
+  self.borderLayer.lineWidth = self.borderPath.lineWidth;
+  self.borderLayer.lineCap = MDCNSStringFromCGLineCap(self.borderPath.lineCapStyle);
+  self.borderLayer.lineJoin = MDCNSStringFromCGLineJoin(self.borderPath.lineJoinStyle);
+  self.borderLayer.miterLimit = self.borderPath.miterLimit;
+  self.borderLayer.path = self.borderPath.CGPath;
+  self.borderLayer.strokeColor = self.borderStrokeColor.CGColor;
+}
+
+- (void)layoutSubviews {
+  [super layoutSubviews];
+
+  [self updateBorder];
+}
+
+#pragma mark - Properties
+
+- (void)setBorderFillColor:(UIColor *)borderFillColor {
+  if (_borderFillColor != borderFillColor) {
```

Comment:
Consider:
if (![_borderFillColor isEqual:borderFillColor]) 

I usually prefer isEqual: for NSObjects.
Here and below.